src/ck3/data/prov_terrain.rs

```diff
@@ -10,32 +9,49 @@ use crate::helpers::dup_error;
 use crate::item::Item;
 use crate::pdxfile::PdxFile;
 use crate::report::{warn, ErrorKey};
-use crate::token::Token;
+use crate::token::{Loc, Token};
 use crate::validator::Validator;
 use crate::Severity;
 
 use super::provinces::ProvId;
 
+const DEFAULT_TERRAINS: &[&str] = &["default_land", "default_sea", "default_coastal_sea"];
+
 #[derive(Clone, Debug, Default)]
 pub struct ProvinceTerrains {
-    provinces: FnvHashSet<ProvinceTerrain>,
+    provinces: FnvHashMap<ProvId, ProvinceTerrain>,
+    file_loc: Option<Loc>,
+    defaults: Box<[Option<Token>; DEFAULT_TERRAINS.len()]>,
 }
 
 impl ProvinceTerrains {
-    fn load_item(&mut self, key: Token, value: Token) {
-        let province = ProvinceTerrain::new(key, value);
-        if let Some(existing_province) = self.provinces.get(&province) {
-            if existing_province.key.loc.kind >= province.key.loc.kind {
-                dup_error(&province.key, &existing_province.key, "province");
+    fn load_item(&mut self, id: ProvId, key: Token, value: Token) {
+        if let Some(province) = self.provinces.get_mut(&id) {
+            if province.key.loc.kind >= key.loc.kind {
+                dup_error(&key, &province.key, "province");
             }
+            province.terrain = value;
```

Comment:
I don't think this is enough, because now the province.key token has the wrong loc.